File: src/main/java/ch/obermuhlner/astro/app/CommandParser.java
package ch.obermuhlner.astro.app;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public class CommandParser {

  private final Map<Command, CommandHandler> commands = new HashMap<>();

  public void add(Command command, CommandHandler handler) {
    commands.put(command, handler);
  }

  public void parse(String[] args) {
    int index = 0;

    while (index < args.length) {
      Command command = findCommand(args[index]);
      if (command == null) {
        throw new RuntimeException("Unknown command: " + args[index]);
      }
      index++;

      List<NamedArguments> optionsWithArguments = new ArrayList<>();
      List<String> commandArguments = new ArrayList<>();

      String optionName = getOptionName(args, index);
      while (optionName != null) {
        index++;

        Option option = command.findOption(optionName);
        if (option == null) {
          throw new RuntimeException("Unknown option for command " + command + ": " + optionName);
        }
        List<String> optionArguments = new ArrayList<>();
        while (optionArguments.size() < option.argumentCount) {
          optionArguments.add(args[index]);
          index++;
        }

        optionsWithArguments.add(new NamedArguments(optionName, optionArguments));

        optionName = getOptionName(args, index);
      }

      while (commandArguments.size() < command.argumentCount) {
        commandArguments.add(args[index]);
        index++;
      }

      CommandHandler commandHandler = commands.get(command);
      commandHandler.handle(new NamedArguments(command.name, commandArguments), new OptionsWithArguments(optionsWithArguments));
    }
  }

  public void foo() {

  }

  private String getOptionName(String[] args, int index) {
    if (args.length <= index) {
      return null;
    }

    String arg = args[index];
    for (String prefix : Arrays.asList("--", "-")) {
      if (arg.startsWith(prefix)) {
        return arg.substring(prefix.length());
      }
    }

    return null;
  }

  private Command findCommand(String arg) {
    for (Command command : commands.keySet()) {
      if (command.name.equals(arg)) {
        return command;
      }
    }

    return null;
  }

  public static class Command {

    private final String name;
    private final int argumentCount;
    private final Option[] options;

    public Command(String name, int argumentCount, Option... options) {
      this.name = name;
      this.argumentCount = argumentCount;
      this.options = options;
    }

    public Option findOption(String optionName) {
      for (Option option : options) {
        if (option.name.equals(optionName)) {
          return option;
        }
      }
      return null;
    }
  }

  public static class Option {

    private final String name;
    private final int argumentCount;

    public Option(String name, int argumentCount) {
      this.name = name;
      this.argumentCount = argumentCount;
    }
  }

  public static class NamedArguments {
    public final String name;
    public final List<String> arguments;

    public NamedArguments(String name, List<String> arguments) {
      this.name = name;
      this.arguments = arguments;
    }

    @Override
    public String toString() {
      return name + " : " + arguments;
    }
  }

  public static class OptionsWithArguments {

    private final List<NamedArguments> optionsWithArguments;

    OptionsWithArguments(List<NamedArguments> optionsWithArguments) {
      this.optionsWithArguments = optionsWithArguments;
    }

    public List<String> getOptionArguments(String optionName, String... defaultArguments) {
      for (NamedArguments optionWithArguments : optionsWithArguments) {
        if (optionWithArguments.name.equals(optionName)) {
          return optionWithArguments.arguments;
        }
      }
      return Arrays.asList(defaultArguments);
    }

    public int handleOption(String optionName, Consumer<List<String>> handleOptionArguments) {
      int count = 0;
      for (NamedArguments optionWithArguments : optionsWithArguments) {
        if (optionWithArguments.name.equals(optionName)) {
          handleOptionArguments.accept(optionWithArguments.arguments);
          count++;
        }
      }
      return count;
    }
  }

  public interface CommandHandler {
    void handle(NamedArguments commandArguments, OptionsWithArguments optionsWithArguments);
  }
}

File: src/main/java/ch/obermuhlner/astro/app/CommandParser.kt
package ch.obermuhlner.astro.app

import java.util.*
import java.util.function.Consumer

class CommandParser {
    private val commands: MutableMap<Command, CommandHandler> = HashMap()
    fun add(command: Command, handler: CommandHandler) {
        commands[command] = handler
    }

    fun parse(args: Array<String>) {
        var index = 0
        while (index < args.size) {
            val command = findCommand(args[index]) ?: throw RuntimeException("Unknown command: " + args[index])
            index++
            val optionsWithArguments: MutableList<NamedArguments> = ArrayList()
            val commandArguments: MutableList<String?> = ArrayList()
            var optionName = getOptionName(args, index)
            while (optionName != null) {
                index++
                val option = command.findOption(optionName)
                        ?: throw RuntimeException("Unknown option for command $command: $optionName")
                val optionArguments: MutableList<String?> = ArrayList()
                while (optionArguments.size < option.argumentCount) {
                    optionArguments.add(args[index])
                    index++
                }
                optionsWithArguments.add(NamedArguments(optionName, optionArguments))
                optionName = getOptionName(args, index)
            }
            while (commandArguments.size < command.argumentCount) {
                commandArguments.add(args[index])
                index++
            }
            val commandHandler = commands[command]
            commandHandler!!.handle(NamedArguments(command.name, commandArguments), OptionsWithArguments(optionsWithArguments))
        }
    }

    fun foo() {}
    private fun getOptionName(args: Array<String>, index: Int): String? {
        if (args.size <= index) {
            return null
        }
        val arg = args[index]
        for (prefix in Arrays.asList("--", "-")) {
            if (arg.startsWith(prefix)) {
                return arg.substring(prefix.length)
            }
        }
        return null
    }

    private fun findCommand(arg: String): Command? {
        for (command in commands.keys) {
            if (command.name == arg) {
                return command
            }
        }
        return null
    }

    class Command(val name: String, val argumentCount: Int, vararg opts: Option) {
        private val options: Array<Option> = opts as Array<Option>
        fun findOption(optionName: String): Option? {
            for (option in options) {
                if (option.name == optionName) {
                    return option
                }
            }
            return null
        }
    }

    class Option(val name: String, val argumentCount: Int)
    class NamedArguments(val name: String, val arguments: List<String?>) {
        override fun toString(): String {
            return "$name : $arguments"
        }
    }

    class OptionsWithArguments internal constructor(private val optionsWithArguments: List<NamedArguments>) {
        fun getOptionArguments(optionName: String, vararg defaultArguments: String?): List<String?> {
            for (optionWithArguments in optionsWithArguments) {
                if (optionWithArguments.name == optionName) {
                    return optionWithArguments.arguments
                }
            }
            return Arrays.asList(*defaultArguments)
        }

        fun handleOption(optionName: String, handleOptionArguments: Consumer<List<String?>>): Int {
            var count = 0
            for (optionWithArguments in optionsWithArguments) {
                if (optionWithArguments.name == optionName) {
                    handleOptionArguments.accept(optionWithArguments.arguments)
                    count++
                }
            }
            return count
        }
    }

    interface CommandHandler {
        fun handle(commandArguments: NamedArguments?, optionsWithArguments: OptionsWithArguments?)
    }
}